### PR TITLE
feat(plugins): Expand issue tracking storage

### PR DESCRIPTION
**Pull request by dcramer**
Add capability to store arbitrary metadata when linking or creating issues.

This is based on code present in sentry-plugins, but abstracted to handle legacy cases better.

### CODE REVIEW

**Comment by dcramer** [comment]
I think this still needs a lot of clean up to make it easier to test/maintain, but this is a first pass at pulling in the code

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/plugins/bases/issue2.py [hunk 1]
```diff
@@ -215,20 +249,27 @@ def view_create(self, request, group, **kwargs):
             return Response({'error_type': 'validation', 'errors': errors}, status=400)
 
         try:
-            issue_id = self.create_issue(
+            issue = self.create_issue(
```
Comment (dcramer) [inline src/sentry/plugins/bases/issue2.py]: actually this likely still isnt legacy compat fully

im going to get some tests in to ensure it doesnt break existing behavior

**Comment by mitsuhiko** [comment]
I looked at the PR and it generally seems to make sense but one test is failing.

**Review by mattrobenolt — APPROVED** [review]
🍪

**Comment by MaxBittker** [comment]
what's the status of this @dcramer ? still want to land it?

**Comment by dcramer** [comment]
I might take a look at cleaning this up to help expedite the mover to new integrations. I reviewed the code again and it MIGHT be stable but still needs tests + qa'd.

File: src/sentry/plugins/bases/issue2.py (hunk 1 above)
Comment (dcramer) [reply src/sentry/plugins/bases/issue2.py]: this comment I believe is addressed, as we change 'issue' below

that said, there's no tests proving this works or doesnt work

**Comment by dcramer** [comment]
Tests are passing again. I'd be content w/ some QA around this and another set of eyes on the diff.

**Comment by dcramer** [comment]
Going to add some unit tests for the various compat methods

**Comment by dcramer** [comment]
@EvanPurkhiser I realize you dont have tons of context on the issue part of the platform, but can I get your thoughts on this? The VSTS plugin does this by reimplementing all these views, and it feels like the right direction to go as a standardized way.

File: src/sentry/plugins/bases/issue2.py [hunk 2]
```diff
@@ -139,31 +145,28 @@ def get_new_issue_fields(self, request, group, event, **kwargs):
...
         e.g. GitHub represents issues as GH-XXX
         """
-        return '#%s' % issue_id
+        if isinstance(issue, dict):
+            return '#{}'.format(issue['id'])
+        return '#{}'.format(issue)
 
     def create_issue(self, request, group, form_data, **kwargs):
         """
         Creates the issue on the remote service and returns an issue ID.
```
Comment (EvanPurkhiser) [inline src/sentry/plugins/bases/issue2.py]: Comment is slightly wrong now, maybe 'issue details'

File: src/sentry/plugins/bases/issue2.py [hunk 3]
```diff
@@ -239,26 +284,25 @@ def view_create(self, request, group, **kwargs):
...
-    def view_unlink(self, request, group, **kwargs):
-        auth_errors = self.check_config_and_auth(request, group)
-        if auth_errors:
-            return Response(auth_errors, status=400)
-        if GroupMeta.objects.get_value(group, '%s:tid' % self.get_conf_key(), None):
-            if 'unlink' in self.allowed_actions:
-                GroupMeta.objects.unset_value(group, '%s:tid' % self.get_conf_key())
-                return Response({'message': 'Successfully unlinked issue.'})
-        return Response({'message': 'No issues to unlink.'}, status=400)
+        return Response({'issue_url': self.get_issue_url(group=group, issue=issue)})
```
Comment (dcramer) [inline src/sentry/plugins/bases/issue2.py]: need to change this to use args instead of kwargs for get_issue_url

**Comment by dcramer** [comment]
going to QA this today and try to get it landed

**Comment by dcramer** [comment]
I think I'm good with this. An important point to note is this i only available for IssueTrackingPlugin2. In retrospect I could have always made an IssueTrackingPlugin3 vs changing this API, but i feel like its not worth it.

**Pull request merged**